core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -313,35 +313,76 @@ class RequestSendThread(val controllerId: Int,
   }
 }
 
-class ControllerBrokerRequestBatch(config: KafkaConfig,
-                                   controllerChannelManager: ControllerChannelManager,
-                                   controllerEventManager: ControllerEventManager,
-                                   controllerContext: ControllerContext,
-                                   stateChangeLogger: StateChangeLogger)
-  extends AbstractControllerBrokerRequestBatch(config, controllerContext, stateChangeLogger) {
+class ControllerBrokerRequestBatch(
+  config: KafkaConfig,
+  controllerChannelManager: ControllerChannelManager,
+  controllerEventManager: ControllerEventManager,
+  controllerContext: ControllerContext,
+  stateChangeLogger: StateChangeLogger
+) extends AbstractControllerBrokerRequestBatch(
+  config,
+  () => controllerContext,
+  () => config.interBrokerProtocolVersion,
+  stateChangeLogger
+) {
 
   def sendEvent(event: ControllerEvent): Unit = {
     controllerEventManager.put(event)
   }
-
   def sendRequest(brokerId: Int,
                   request: AbstractControlRequest.Builder[_ <: AbstractControlRequest],
                   callback: AbstractResponse => Unit = null): Unit = {
     controllerChannelManager.sendRequest(brokerId, request, callback)
   }
 
+  override def handleLeaderAndIsrResponse(response: LeaderAndIsrResponse, broker: Int): Unit = {
+    sendEvent(LeaderAndIsrResponseReceived(response, broker))
+  }
+
+  override def handleUpdateMetadataResponse(response: UpdateMetadataResponse, broker: Int): Unit = {
+    sendEvent(UpdateMetadataResponseReceived(response, broker))
+  }
+
+  override def handleStopReplicaResponse(stopReplicaResponse: StopReplicaResponse, brokerId: Int,
+                                         partitionErrorsForDeletingTopics: Map[TopicPartition, Errors]): Unit = {
+    if (partitionErrorsForDeletingTopics.nonEmpty)
+      sendEvent(TopicDeletionStopReplicaResponseReceived(brokerId, stopReplicaResponse.error, partitionErrorsForDeletingTopics))
+  }
+}
+
+abstract class ControllerBrokerRequestMetadata {
+  def isTopicDeletionInProgress(topicName: String): Boolean
+
+  def topicIds: collection.Map[String, Uuid]
+
+  def liveBrokerIdAndEpochs: collection.Map[Int, Long]
+
+  def liveOrShuttingDownBrokers: collection.Set[Broker]
+
+  def isTopicQueuedUpForDeletion(topic: String): Boolean
+
+  def isReplicaOnline(brokerId: Int, partition: TopicPartition): Boolean
+
+  def partitionReplicaAssignment(partition: TopicPartition): collection.Seq[Int]
+
+  def leaderEpoch(topicPartition: TopicPartition): Int
+
+  def liveOrShuttingDownBrokerIds: collection.Set[Int]
+
+  def partitionLeadershipInfo(topicPartition: TopicPartition): Option[LeaderIsrAndControllerEpoch]
 }
 
 abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
```

Comment:
It might be worthwhile to document the arguments now that we have a few that vary between KRaft and ZK modes.